src/openassetio-python/private/include/openassetio/private/python/pointers.hpp

```diff
@@ -48,10 +49,19 @@ Ptr createPyRetainingPtr(const py::object& pyInstance,
   // of the original shared_ptr holder stored on the PyObject.
   auto pyInstancePtr =
       std::shared_ptr<py::object>{new py::object{pyInstance}, [](py::object* pyObjectPtr) {
-                                    // Acquire the GIL, in case deleter
-                                    // runs in a non-Python thread.
-                                    py::gil_scoped_acquire gil;
-                                    delete pyObjectPtr;
+                                    if (_Py_IsFinalizing()) {
+                                      // If the Python interpreter is
+                                      // gone, clear the internal
+                                      // PyObject* so pybind11 won't
+                                      // attempt to clean it up.
+                                      pyObjectPtr->release();
+                                      delete pyObjectPtr;
+                                    } else {
+                                      // Acquire the GIL, in case deleter
+                                      // runs in a non-Python thread.
+                                      py::gil_scoped_acquire gil;
```

Comment:
[question] Asking to make sure we've considered it: What are the failure modes of this call, and do we have to worry about them happening in a destructor?

Reply:
Looking at the code of `gil_scoped_acquire`, it's mostly calls to CPython. The only exceptions that can be thrown are catastrophic errors with CPython that shouldn't ever occur, as far as I can tell.

Reply:
Ace, thanks for the extra due dill.

Reply:
Not a bad idea to wrap in a `try`/`catch`, though :thinking:

Reply:
That sounds more important than the try/catch? Given your notes on "the general badness that would be happening at the same time anyway"?